src/spectral.jl

```diff
@@ -133,3 +133,33 @@ function CombinatorialAdjacency(g::Graph)
     return CombinatorialAdjacency{Float64, typeof(g), typeof(d)}(g,d)
 end
 end # @require
+
+
+"""Returns a sparse node-arc incidence matrix for a graph, indexed by
+`[v, i]`, where `i` is in `1:ne(g)`, indexing an edge `e`. For
+directed graphs, a value of `-1` indicates that `src(e) == v`, while a
+value of `1` indicates that `dst(e) == v`. Otherwise, the value is
+`0`. For undirected graphs, both entries are `1`.
+"""
+function incidence_matrix(g::SimpleGraph, T::DataType=Int)
+    n_v = nv(g)
+    n_e = ne(g)
+    nz = 2 * n_e
+
+    # every col has the same 2 entries
+    u_coef = is_directed(g)? -one(T) : one(T)
+    v_coef = one(T)
+    colpt = collect(1:2:(nz + 1))
+    nzval = repmat([u_coef, v_coef], n_e)
+
+    # iterate over edges for row indices
+    rowval = zeros(Int, nz)
+    for (i,e) in enumerate(edges(g))
```

Comment:
so you're still iterating over edges here - this is likely to be give poor performance. Can you try `@time` with medium-sized (1e6,1e7) graphs?